mp4parse_capi/src/lib.rs

```diff
@@ -417,76 +417,66 @@ impl Read for Mp4parseIo {
-/// a valid Mp4parseStatus.
-///
-/// On a successful the variable pointed to by `status_out` will be set to
-/// `Mp4parseStatus::Ok` and the return value will be a non-NULL pointer to
-/// an `Mp4parseParser`. If there is any sort of error, the return value will
-/// be NULL and the type of error will be set in the variable pointed to by
-/// `status_out`, ensuring the consumer never has access to potentially invalid
-/// parsed data.
+/// This function is unsafe because it dereferences the `io` and `parser_out`
+/// pointers given to it. The caller should ensure that the `Mp4ParseIo`
+/// struct passed in is a valid pointer. The caller should also ensure the
+/// members of io are valid: the `read` function should be sanely implemented,
+/// and the `userdata` pointer should be valid. The `parser_out` should be a
+/// valid pointer to a location containing a null pointer. Upon successful
+/// return (`Mp4parseStatus::Ok`), that location will contain the address of
+/// an `Mp4parseParser` allocated by this function.
 ///
 /// To avoid leaking memory, any successful return of this function must be
 /// paired with a call to `mp4parse_free`. In the event of error, no memory
 /// will be allocated and `mp4parse_free` must *not* be called.
 #[no_mangle]
 pub unsafe extern "C" fn mp4parse_new(
     io: *const Mp4parseIo,
-    status_out: *mut Mp4parseStatus,
-) -> *mut Mp4parseParser {
-    mp4parse_new_common(io, status_out)
+    parser_out: *mut *mut Mp4parseParser,
+) -> Mp4parseStatus {
+    mp4parse_new_common(io, parser_out)
 }
 
 /// Allocate an `Mp4parseAvifParser*` to read from the supplied `Mp4parseIo`.
 ///
-/// See mp4parse_new; this function is identical except that it returns a
-/// pointer to an `Mp4parseAvifParser`, which (when non-null) must be paired
-/// with a call to mp4parse_avif_free.
+/// See mp4parse_new; this function is identical except that it allocates an
+/// `Mp4parseAvifParser`, which (when successful) must be paired with a call
+/// to mp4parse_avif_free.
 ///
 /// # Safety
 ///
 /// Same as mp4parse_new.
 #[no_mangle]
 pub unsafe extern "C" fn mp4parse_avif_new(
     io: *const Mp4parseIo,
-    status_out: *mut Mp4parseStatus,
-) -> *mut Mp4parseAvifParser {
-    mp4parse_new_common(io, status_out)
+    parser_out: *mut *mut Mp4parseAvifParser,
+) -> Mp4parseStatus {
+    mp4parse_new_common(io, parser_out)
 }
 
 unsafe fn mp4parse_new_common<P: ContextParser>(
     io: *const Mp4parseIo,
-    status_out: *mut Mp4parseStatus,
-) -> *mut P {
+    parser_out: *mut *mut P,
+) -> Mp4parseStatus {
     // Validate arguments from C.
-    if io.is_null() || (*io).userdata.is_null() || (*io).read.is_none() || status_out.is_null() {
-        set_if_non_null(status_out, Mp4parseStatus::BadArg);
-        std::ptr::null_mut()
+    if io.is_null()
+        || (*io).userdata.is_null()
+        || (*io).read.is_none()
+        || parser_out.is_null()
+        || !(*parser_out).is_null()
```

Comment:
This check requires the caller to null-initialize the outparam before passing in the address, but it's fairly common in C languages to use the address of an uninitialized variable for outparams, e.g.

```c
Mp4ParseParser* parser;
int rv = mp4parse_new(&io, &parser);
```

I think it's better to permit this common pattern than to enforce pre-initialization, but as you've also added a test for if you'd prefer to keep this as-is, the documentation for mp4parse_{avif,}_new should mention it.

Reply:
My thinking here was that there's no way to tell whether a non-NULL passed-in value was merely uninitialized, or a valid pointer which should've been freed. If it's the latter case, calling this function will overwrite the value, possibly leaking. So, the value of possibly catching a leak early seemed to outweigh the cost of requiring callers to initialize to NULL.

That said, I did intend to document it, so thanks for catching that. I'll update shortly.